query-engine/connectors/mongodb-query-connector/src/root_queries/update/expression.rs

```diff
@@ -0,0 +1,288 @@
+            cond: Box::new(cond.into()),
+            then: Box::new(then.into()),
+            els: Box::new(els.into()),
+        })
+    }
+
+    pub(crate) fn try_into_set(self) -> Option<Set> {
+        if let Self::Set(v) = self {
+            Some(v)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn as_merge_objects_mut(&mut self) -> Option<&mut MergeObjects> {
+        if let Self::MergeObjects(ref mut v) = self {
+            Some(v)
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct Set {
+    /// The field path to which this $set expression should be applied
+    pub field_path: FieldPath,
+    /// The inner expression that should be set
+    pub expression: Box<UpdateExpression>,
+}
+
+impl Set {
+    /// Get a reference to the set's field path.
+    pub(crate) fn field_path(&self) -> &FieldPath {
+        &self.field_path
+    }
+
+    /// Transforms a `Set` expression into a conditional one. eg:
+    /// ```text
+    /// from: { $set: { {set.field_path}: {set.expression} } }
+    /// to:   { $set: { $cond: { if: {cond}, then: {set.expression}, else: "${set.field_path}"  } } }
+    /// ```
+    pub(crate) fn into_conditional_set(self, cond: impl Into<UpdateExpression>) -> Self {
+        let dollar_path = self.field_path().dollar_path(true);
+
+        Self {
+            field_path: self.field_path,
+            expression: Box::new(UpdateExpression::IfThenElse(IfThenElse {
+                cond: Box::new(cond.into()),
+                then: self.expression,
+                els: Box::new(UpdateExpression::Generic(Bson::String(dollar_path))),
+            })),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct IfThenElse {
+    /// The condition of the if expression
+    pub cond: Box<UpdateExpression>,
+    /// The then branch
+    pub then: Box<UpdateExpression>,
+    /// The else branch
+    pub els: Box<UpdateExpression>,
+}
+
+#[derive(Debug, Clone, Default)]
+pub(crate) struct MergeDocument {
+    pub inner: IndexMap<String, UpdateExpression>,
+}
+
+impl IntoIterator for MergeDocument {
+    type Item = (String, UpdateExpression);
+    type IntoIter = indexmap::map::IntoIter<String, UpdateExpression>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.inner.into_iter()
+    }
+}
+
+impl MergeDocument {
+    /// Get a mutable reference to the merge document's inner.
+    pub(crate) fn inner_mut(&mut self) -> &mut IndexMap<String, UpdateExpression> {
+        &mut self.inner
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct MergeObjects {
+    /// The field path to which this $mergeObjects expression should be applied
+    /// { $mergeObjects: [<field_path>, ...] }
+    pub(crate) field_path: FieldPath,
+    /// Document against which it should be merged
+    /// { $mergeObjects: [..., <document>] }
+    pub(crate) document: MergeDocument,
+    /// Hack: keys of the document located in `field_path` to unset
```

Comment:
This is a hack because `MergeObject` (just like any other `UpdateExpression`) is supposed to be a generic structure that represents a `$mergeObjects` expression in MongoDB. However, `keys_to_unset` is something completely made up and very specific to our needs so that we can easily render `unset` in a `$mergeObjects`.

Ideally, instead of taking a `FieldPath` for the first argument of `$mergeObjects: [<the field path>,  ...]`, it should:
- take an `UpdateExpression` which could either be a `MergeDocument` or a `FieldPath`
- Fill the `MergeObjects` structure with whatever are our needs during the `Operation` to `Expression` layer: either a `FieldPath` or the `$arrayToObject + $filter + $objectToArray` crap to unset fields
- Leave the `Expression` to `Bson` layer free from any domain-specific needs